### DISCUSSION TITLE
Revising Project Hierarchy within the Fluent Organization

### DISCUSSION THREAD

**Discussion by edsiper** [fluent/fluentd#4471] [General]
Dear Fluent community, 

the following is an open discussion about the upcoming proposal to switch Fluent Bit as the primary project in the Fluent organization for CNCF. Feel free to jump in and share your thoughts. After feedback and adjustments, we will run an open voting process. This is the first step.

## Introduction

The [Fluent](https://github.com/fluent) organization has been pivotal in addressing the complexities of Telemetry Data management across infrastructure by enabling reliable collection, processing, and delivery, while integrating seamlessly with other ecosystems. Since its inception, [Fluentd](https://github.com/fluent/fluentd), our original primary project, has significantly contributed to this mission. Established in 2011 and donated to the [Cloud Native Computing Foundation (CNCF)](https://cncf.io) in 2016, Fluentd achieved CNCF graduation in 2019. This milestone marked its widespread adoption by leading cloud providers such as AWS, Google, and Microsoft, solidifying its role in enterprise logging.

## Evolution of Technology and User Demands

As technology and user expectations have evolved, so has the demand for high-performance solutions capable of handling diverse types of Telemetry data,  [Fluent Bit](https://github.com/fluent/fluent-bit), created in 2015,  emerged as a response to these needs. Designed to be lightweight and efficient, Fluent Bit offers enhanced performance and the capability to manage not only logs but also metrics and traces. This adaptability has led to its increased adoption, making it a preferred choice for many users, and has positioned it at the forefront of innovation within our organization.

## Proposal for Organizational Change

Given the strategic and technological shift towards Fluent Bit, there is a strong case for reevaluating its position within the Fluent organization. This proposal suggests considering [Fluent Bit](https://github.com/fluent/fluent-bit) as the parent project due to its broader capabilities and active development focus. Consequently, [Fluentd](https://github.com/fluent/fluentd) would transition to a role that reflects its current state of being in a maintenance phase, acknowledging its continued importance but recognizing the shift in innovation and user focus towards [Fluent Bit](https://github.com/fluent/fluent-bit).

## Next steps proposed

- Update the GOVERNANCE files for Fluentd and Fluent Bit.
- Primary project for Fluent organization be Fluent Bit.
  - Change the organization logo from Fluentd to Fluent Bit.
- Open the voting process once PRs for Governance are clear.

### Why Fluent Bit ?

__Adoption and Performance__: With over 13 billion downloads in containerized environments alone, Fluent Bit demonstrates high adoption rates and proven performance. This extensive usage underscores its reliability and efficiency in handling telemetry data at scale.

__Cloud Provider Endorsement__: Major cloud providers, including AWS, Google, and Microsoft, have transitioned from using Fluentd to Fluent Bit for their logging and telemetry needs. This shift is largely due to Fluent Bit's superior performance, lower resource consumption, and its ability to handle high-volume data more efficiently.

__Versatility in Data Handling__: Unlike Fluentd, which primarily focuses on log data, Fluent Bit is designed to manage multiple types of telemetry data, including logs, metrics, and traces. This versatility makes Fluent Bit more suited to current and future demands of complex data pipelines and observability frameworks.

__Active Development and Innovation__: Fluent Bit enjoys a vibrant development community that regularly introduces new features and enhancements. This active innovation pipeline has kept the project at the cutting edge, adapting quickly to new technologies and user requirements.

### What's next for Fluentd ?

Fluentd is still widely deployed, and our organization's goal is to maintain it for the long term. If some company or individual desires to contribute innovation in the project in the future, it is totally possible. Releases and development will **not stop**.

### Are there any License changes involve ?

__NO__. Both, Fluentd and Fluent are open source projects hosted by the [Cloud Native Computing Foundation](https://cncf.io) under the terms of the Apache v2 License. We are vendor-neutral projects.

**Comment by daipom** [comment]
Thank you for your proposal for the future of the Fluent organization.
I would be happy to make any improvements that would be good for the whole Fluent organization.

I understand the background and the situation, but I have a question about the way to change the primary project.
My question is:

* Wouldn't it be better to start with a form in which they are aligned in parallel?

First of all, we should clarify what the `primary project` is.
In reality, I don't think it is a situation where one of them is Primary and one of them is a Sub.

I think there are the following two points.
Is it correct?

* `A`: The title of Fluent organization: https://github.com/fluent
  * Since this is completely Fluentd content now, it should be improved somehow.
* `B`: CNCF's project: https://www.cncf.io/projects/
  * As an Observability CNCF project, there is very little information about Fluent Bit, so it should be improved somehow.

As for `A`, the issue of what to do with the logo would remain, but other points for `A` and `B` could be greatly improved by changing the content so that Fluentd and Fluent Bit are aligned in parallel.

* `A`: Change the title so that it contains both contents for Fluentd and Fluent Bit.
* `B`: Get Fluent Bit lined up with Fluentd as one Observability project, etc.

I think it's better to start by considering improvements that would add Fluent Bit information as much as Fluentd, rather than simply reversing the current state.
How about first considering improvements that will get Fluentd and Fluent Bit aligned in parallel?

**Reply by cosmo0920** [reply]
> `B`: CNCF's project: https://www.cncf.io/projects/
> 
>     * As an Observability CNCF project, Fluent Bit is not as outstanding as Fluentd, so it should be improved somehow.

Currently, Fluentd does not have capabilities to handle metrics and traces in their pipelines.  Fluent Bit already _does_ have capabilities to handle metrics and traces like as I described/created in the latest Forward protocol v1.5.
CNCF's page describes Fluentd as a one of the observability projects. Although, it does not have capabilities to handle prometheus compatible payloads or opentelemetry traces.
For cloud native generation of the architectures, users really need to handle observabilities because many of containers are ephemeral and should be prepared for orchestration tool for fleet management.

In this context, we need to emphasize that the observability supported agent which is Fluent Bit.

In my understanding, this primary project switch is just like: Which agent should be emphasized in the current observability context? Cloud native architecture is needed to adopt lightweight but be able to handle different types of events.

For competition perspective,  Prometheus ecosystem has metrics capability as Prometheus itself and Prometheus Loki can handle logs and Grafana Tempo which handles traces in their ecosystem.
Also, [Jeager](https://www.jaegertracing.io/) emphasizes to analyze traces capabilities.
So, we need to emphasize that we have already an observability agent somehow.

In our situation, it is a bit of complex due to Fluentd has long standing project but it normally handles for logs and pertially handles metrics via fluent-plugin-prometheus.
By contrast, Fluent Bit has builtin features for metrics and traces. Also, it can handle relatively large amount of logs.

> I think it's better to start by considering improvements that would make Fluent Bit as outstanding as Fluentd, rather than simply reversing the current state.
> How about first considering improvements that will give Fluentd and Fluent Bit equal outstanding?

How come the "equal outstanding"?
We needn't to improve as an equal project. Because Fluent Bit is lightweight and implementing as the next generation of unified layer. This could be included the meanings of obserbavilities which are logs, metrics, and traces.

**Reply by daipom** [reply]
> How come the "equal outstanding"?
We needn't to improve as an equal project. Because Fluent Bit is lightweight and implementing as the next generation of unified layer. This could be included the meanings of obserbavilities which are logs, metrics, and traces.

Sorry, maybe my English was inappropriate.

I don't mean that Fluent Bit and Fluentd should have equal capabilities.
I'm saying that it would be nice to be able to arrange Fluent Bit information as much as Fluentd information in https://github.com/fluent and https://www.cncf.io/projects/ since the current information is mostly for Fluentd.

(outstandingを「目立つ」という意図で使いましたが、「機能が優れている」という意味になってしまいますかね。すみません、私の英語のミスです。)

**Reply by cosmo0920** [reply]
Got it. It should be aligned for equally arranged for Fluentd and Fluent Bit in the web page and project page.

**Reply by daipom** [reply]
I was using the word "outstanding" with the wrong meaning. Sorry for the confusion.
I have fixed some sentences.

**Reply by cosmo0920** [reply]
However, even if equally displaying the information of Fluentd and Fluent Bit as much as possible, the seat of the project page in CNCF could be only one. So, we need to put the most preferred project that the future users could use and try at first. :)

**Reply by daipom** [reply]
>  the seat of the project page in CNCF could be only one.

I'm concerned about this point.
I am sorry that I have not been able to confirm this in detail, but I can find multiple Observability projects, so, for example, I wonder if we could add Fluent Bit as an independent CNCF Observability project.

**Reply by edsiper** [reply]
@daipom thanks for your feedback. 

Actually there is no harm for Fluentd when doing this change, and is all positive for Fluent Bit. What we are proposing is just to formalize where the industry is going. All major cloud providers already migrated and many other companies as well. 

Adding Fluent Bit as an independent project is a step backward, we are with CNCF since the early years and it was a ton of work (years) to get the ecosystem in, initially as incubated and then handling the process for graduation; starting as a new project means to start all over again, which I think there is no value for the project. 

As mentioned the goal is to switch the focus at a CNCF level for exposure.

**Reply by kenhys** [reply]
It may be reasonable to focus on Fluent Bit in fluent organization to recommend.

As a result, it may be changed like:

* https://github.com/fluent
  * Change Logo to Fluent-Bit may be good for making it attractive. No objection at all.
* https://www.cncf.io/projects/fluentd/ 
  * Adding explanation about Fluent Bit may be good for making it attractive.

But not changed listed in:

* https://www.cncf.io/projects/
  * Even though Fluentd was changed to sub-project position, Fluentd is still shown in this page, isn't  it? I
  think so because it depends on graduated project or not.

If you mean "the seat of the project page in CNCF"  is listing in Fluent Bit as observability https://www.cncf.io/projects/
as a result of switching primary - sub-project relationship between Fluentd and Fluent Bit,  it is not proper way to go IMHO.

Thus, it is inappropriate to change the primary - subproject relationship in a project just to avoid formal procedures especially not good for Fluent Bit itself. (I'm worried about it gives third-party bad impressions)

(Personally, I felt that it's time to revisit project hierarchy not the umbrella under Fluentd, but also to  make Fluent Bit as an another top project in CNCF. e.g like helm graduated from kubernetes project.)


FYI: https://helm.sh/docs/community/history/

> Helm was promoted from a Kubernetes subproject to a full-fledged CNCF project in June, 2018.

**Reply by edsiper** [reply]
The primary intention is not to skip formal procedures, the project is already there, but switch focus to reality. I cannot remember how many people migrated out of Fluentd because of performance needs without them knowing that Fluent Bit is there. On the other side, Fluent Bit is deployed a million times per month. This is very confusing for CNCF community in general.

I would consider Helm a different case, since that's a totally different project, here Fluent Bit is kind of the next-gen of Fluent.

**Reply by edsiper** [reply]
FYI: I have shared this feedback with the CNCF TOC and they have provided good guidance. I will create another discussion with a different approach to continue there.

**Reply by edsiper** [reply]
I have opened the new discussion here:

https://github.com/fluent/fluentd/discussions/4702

I am closing this discussion since the original idea (this discussion) won't move forward based on feedback from maintainers and CNCF.
